eclair-core/src/main/scala/fr/acinq/eclair/channel/TxPublisher.scala

```diff
@@ -239,192 +144,28 @@ private class TxPublisher(nodeParams: NodeParams, watcher: ActorRef[ZmqWatcher.C
 
       case WrappedCurrentBlockCount(blockCount) =>
         val toPublish = cltvDelayedTxs.view.filterKeys(_ <= blockCount)
-        toPublish.values.flatten.foreach(tx => publish(tx))
-        run(cltvDelayedTxs -- toPublish.keys, csvDelayedTxs)
-
-      case PublishNextBlock(p) =>
-        val nextBlockCount = nodeParams.currentBlockHeight + 1
-        val cltvDelayedTxs1 = cltvDelayedTxs + (nextBlockCount -> (cltvDelayedTxs.getOrElse(nextBlockCount, Seq.empty) :+ p))
-        run(cltvDelayedTxs1, csvDelayedTxs)
-
-      case SetChannelId(remoteNodeId, channelId) =>
-        Behaviors.withMdc(Logs.mdc(remoteNodeId_opt = Some(remoteNodeId), channelId_opt = Some(channelId))) {
-          run(cltvDelayedTxs, csvDelayedTxs)
-        }
-    }
-
-  private def publish(p: PublishTx): Future[ByteVector32] = {
-    p match {
-      case SignAndPublishTx(txInfo, commitments) =>
-        log.info("publishing {}: input={}:{} txid={} tx={}", txInfo.desc, txInfo.input.outPoint.txid, txInfo.input.outPoint.index, p.tx.txid, p.tx)
-        val publishF = txInfo match {
-          case tx: ClaimLocalAnchorOutputTx => publishLocalAnchorTx(tx, commitments)
-          case tx: HtlcTx => publishHtlcTx(tx, commitments)
+        toPublish.values.flatten.foreach(tx => publish(tx, channelInfo))
+        run(cltvDelayedTxs -- toPublish.keys, csvDelayedTxs, channelInfo)
+
+      case WrappedTxPublishResult(result) =>
+        result match {
+          case _: TxPublish.InsufficientFunds =>
+            // We retry when the next block has been found, we may have more funds available.
+            val nextBlockCount = nodeParams.currentBlockHeight + 1
+            val cltvDelayedTxs1 = cltvDelayedTxs + (nextBlockCount -> (cltvDelayedTxs.getOrElse(nextBlockCount, Seq.empty) :+ result.cmd))
```

Comment:
That's a bit hacky, the transaction isn't really cltv-delayed here. Also, there is a potential risk of creating a herd effect if our wallet balance is low. That would be a nice stress test of our ability to concurrently fund many transactions and manages locks, but I'm not sure it's a good thing.

Reply:
That will be improved later, when we'll introduce deadlines (block heights at which we should republish/bump some txs), but that requires more work. In the meantime I'd rather err on the safety side and republish aggressively to ensure we don't forget to publish an important transaction because of an unknown error.